### PR TITLE
Bugfix/placing second order

### LINKED ISSUE #2195: "Error placing an order" on Checkout
## Current behavior
while placing an order, Magento 2 looks like quote sometimes is not saved/updated at that moment. As result API returns 500 error in `api/order/create` response "Error: Error placing an order"

http://demo.vuestorefront.io

screen shot 2019-01-15 at 10 33 34 am

screen shot 2019-01-15 at 10 33 56 am

## Steps to reproduce the issue
Add product to cart, directly proceed to checkout and click on "Place order" button. 

## Repository
master branch

## Can you handle fixing this bug by yourself?
no

## Environment details

- Browser: Chrome 71
- OS: MacOS Mojave 10.14.2
- Node: v8.12.0

## Additional information

**API logs:**
```
debug: Response received.
GET /api/cart/payment-methods?token=&cartId=Pw8pGGQ1z1GfIiv6NNbtmaTJlJzOtEmi 200 3481.787 ms - 410
debug: Response received.
error: API call failed: Shipping method is not applicable for empty cart
POST /api/cart/shipping-information?token=&cartId=Pw8pGGQ1z1GfIiv6NNbtmaTJlJzOtEmi 500 3832.740 ms - 72
debug: Response received.
GET /api/stock/check?sku=100759 200 2050.692 ms - 644
debug: Response received.
POST /api/cart/update?token=&cartId=Pw8pGGQ1z1GfIiv6NNbtmaTJlJzOtEmi 200 5066.396 ms - 160
```
Cannot see any logs in Magento 2 instance connected to this error.

There is similar problem that is connected to current behavior - shipping methods are not loaded and default "DPD Courier" id displayed. After returning to non-checkout page and proceeding to Checkout - default shipping method is replaced with methods from Magento.

Local environment:

screen shot 2019-01-15 at 1 19 06 pm

screen shot 2019-01-15 at 1 19 23 pm

`"directBackendSync": true` in VS and `"useServerQueue": false` in API

### LINKED ISSUE #2503: Cannot place second order
## Current behavior
On demo.vuestorefront.io the first order goes through as expected but the second "place the order" does not work. The button click has no effect. Playing around with the debugger I saw that the second time "place order" function in afterRegisteration.ts is never called 

## Steps to reproduce the issue
we tried on both demo.vuestorefront.com and running the latest master locally

# Can you handle fixing this bug by yourself?

NO

## Environment details

- Browser: Chrome
- OS: Ubuntu and MacOS
- Node: LTS

### PR DESCRIPTION
**Pull request by patzick**
### Related issues

closes #2195 
closes #2503 

### Short description and why it's useful

Checkouts.js has been unbinding all payment methods and caused problem with placing second order.

### Upgrade Notes and Changelog

- [x] No upgrade steps required (100% backward compatibility)
- [x] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/doc/Upgrade%20notes.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing VS sites with this new feature

### CODE REVIEW

**Comment by pkarw** [comment]
I think it’s a good moment to implement this one: https://github.com/DivanteLtd/vue-storefront/issues/2409

**Comment by pkarw** [comment]
#2409 should go to 1.8.3 as a order in order/placeOrder action I believe - to handle not only this issue but as a general solution for passing over the same order multiple times. You can use the same hashing algorithm we’re using for cache (in searchAdapter)

**Comment by pkarw** [comment]
Then, for 1.9.rc as a part of theme 2.0 works please go on with https://github.com/DivanteLtd/vue-storefront/issues/2407

I see this one as a TS, typed, interface based approach without the event bus involved. Rather hooks like we have with the modules.

In 1.9 current order flow should still work but the new refactored one should become default with all core payment methods refactored to it.

**Comment by pkarw** [comment]
@ArturDivante pleaae make sure after @patzick changes that this will be well tested: we should test all the payment methods available, going back to catalog (after order placed), placing another one, checking if there are any console errors or duplicated /api/order requests.

@patzick please do install the PayPal and stripe on test to let Artur test it as well 

Sorry for being so meticulous about this issue but If something is related to orders must be treated with highest care :-)

**Comment by patzick** [comment]
@pkarw so i've changed base to develop. Now we also have changes from master which hasn't been merged to develop yet.

**Comment by pkarw** [comment]
@patzick ok but how does this develop branch merge answers my comments ?:-)

**Comment by patzick** [comment]
@pkarw we have multiple communication channels, so when i responded to email (which i read first) i thought you wouldn't expect same answer here :)
Yes - we should make hooks, but this is a critical bug, so rewriting important part of core to new API for fix doesn't look like safe solution.

I've added Critical label and 1.8.3 milestone to #2409 so i will add this check in other PR.

I will add paypal and stripe plugins to test after merge.

File: src/modules/payment-backend-methods/hooks/afterRegistration.ts [hunk 1]
```diff
@@ -4,8 +4,8 @@ export function afterRegistration({ Vue, config, store, isServer }) {
 
   let correctPaymentMethod = false
 
-  // Place the order. The payload is empty as we don't have any specific info to add for this payment method '{}'
-  const placeOrder = function () {
+  // Place the order. Payload is empty as we don't have any specific info to add for this payment method '{}'
+  const placeOrder = () => {
```
Comment (filrak) [inline src/modules/payment-backend-methods/hooks/afterRegistration.ts]: XD

**Comment by pkarw** [comment]
Thanks! We can merge this one in then

**Pull request merged**